### PR TITLE
DONT REVIEW [perf] Configure chunked loss by token length

### PR DESCRIPTION
**Pull request by felipemello1**
## Summary

- Replace `ChunkedLossWrapper.Config.num_chunks: int = 8` with `chunk_size: int = 2048` (max local tokens per lm_head + loss chunk).
- The sequence is split into `ceil(T / chunk_size)` equal chunks; `T` must be divisible by that count, otherwise the loss raises. This holds for every shipped config (all microbatches are multiples of 2048).
- Inputs that fit in one chunk use the plain `loss_fn(lm_head(hidden))` path.
- `GradAccumulator` tracks only the next write offset.
- Recipes that set the chunk count explicitly now use the default.

## Why

Large loss chunks cause CUDA memory allocation retries. Their logits push peak memory to the card's ceiling, and every retry (`memory/num_alloc_retries`) frees the allocator cache and synchronizes; that, not the loss arithmetic, is the slowdown (in an isolated single-GPU loss benchmark a 40,960-token chunk is as fast as an 8,192-token one). `num_chunks` makes the chunk scale with the microbatch, 512 tokens for a 4k-token microbatch and 40,960 for a 327k-token one, so large microbatches always hit this. A fixed `chunk_size` bounds loss memory to ~13 bytes × chunk_size × vocab_per_rank regardless of microbatch length.

Qwen3-14B, 8×H100, TP=4/DP=2, 327,680 tokens per DP rank, 5 steps, steps 2-5 averaged:

```text
                              TPS     MFU      peak active   alloc retries
main, num_chunks=8 (x2)      3094    32.56%    90.81 GiB      28
chunk_size=40,960            3085    32.47%    90.81 GiB      28   (= num_chunks=8)
chunk_size=8,192             3576    37.63%    83.56 GiB       0   (+15.6% TPS)
```

## Effect of chunk_size

Microbenchmark of the loss alone: a standalone bf16 `nn.Linear(D, V)` + cross entropy on random hidden states, one H100, eager, fwd+bwd, no transformer. Shapes copy two lm_heads. Time flattens above ~4k tokens; memory grows linearly with `chunk_size × V`:

```text
D=5120, V=37,984 (Qwen3-14B lm_head, TP=4 shard)      D=4096, V=248,320 (Qwen3.5 lm_head, no TP)
T = 327,680 tokens                                     T = 16,384 tokens
chunk_size   loss fwd+bwd   peak                        chunk_size   loss fwd+bwd   peak
   512        1084 ms      13.3 GiB                       2,048       221 ms       10.9 GiB   <- default
 1,024         898 ms      13.4 GiB                       4,096       211 ms       17.5 GiB
 2,048         825 ms      13.4 GiB   <- default          8,192       208 ms       30.8 GiB
 4,096         787 ms      13.6 GiB                      16,384       212 ms       47.6 GiB
 8,192         770 ms      14.2 GiB
40,960         757 ms      30.5 GiB   (= num_chunks=8)
```

Default 2048: it is today's effective chunk for the standard 16,384-token pretraining microbatch, so pretraining memory is unchanged, and it is within 7% of the fastest setting on the loss portion (~1% end to end). 8192 would triple the loss peak for large-vocab, no-TP models.

## Validation

```bash
python -m pytest tests/unit_tests/cpu/test_loss.py -k "GradAccumulator or (ChunkedLossWrapper and not SPMD)" \
    torchtitan/experiments/graph_trainer/tests/test_chunked_loss.py
python -m pytest torchtitan/experiments/graph_trainer/tests/test_trace_module.py -k chunked \
    torchtitan/experiments/graph_trainer/tests/test_passes.py -k seq_chunk_marker \
    torchtitan/experiments/graph_trainer/tests/test_graph_pp_runner.py -k chunked
python -m pytest tests/unit_tests/cpu/test_no_new_cli_options.py
```

- `TestChunkedLossWrapperSPMD::test_spmd_matches_eager_and_types` fails identically on main (`_LossParallelCrossEntropy` not registered for SPMD type checking); unrelated.
- Integration debug models (≤ 2,048 tokens) go from 8 chunks to 1, changing summation order; the four affected goldens are regenerated from CI (largest drift 4e-5 relative). The RL loss guard golden (`rl_grpo_cuda.txt`) needs regenerating on the A10G runner after this change.

## Follow-ups

- `GradAccumulator` keeps an fp32 `[T, D]` buffer although chunk grads are bf16 and writes are disjoint; using `hidden_states.dtype` halves it.
- `compile.components=["loss"]` cuts the loss portion ~20% at any chunk size.

### CODE REVIEW

File: docs/chunked_loss.md
Comment (tianyu-l) [inline docs/chunked_loss.md:1]: please put close to where the component is defined

File: docs/assets/chunked_loss_chunk_len_benchmark.svg
Comment (tianyu-l) [inline docs/assets/chunked_loss_chunk_len_benchmark.svg:1]: hmm where does flop reduction come from? Is it because the 8-chunk memory usage is close to memory limit so there're many cuda memory realloc?

File: torchtitan/components/loss.py
Comment (tianyu-l) [inline torchtitan/components/loss.py]: there seem to be some redundancy in `num_chunks`, `_next_idx`, `_next_start` -- it seems to me that in this `GradAccumulator` we shouldn't care about `num_chunks` at all -- we only need to keep `_next_start` and this check can be done by comparing `_next_start`  with buffer / local shape.

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -602,26 +620,45 @@ def __call__(
...
-        # writes, which use one chunk length for each write offset.
+        def _chunk_count(seq_len: int | torch.SymInt) -> int:
+            from torch.fx.experimental.symbolic_shapes import optimization_hint
+
+            seq_len_hint = optimization_hint(seq_len)
+            return (seq_len_hint + self.chunk_len - 1) // self.chunk_len
+
         def _chunk_local(t):
             seq_len = t.shape[0]
+            if isinstance(seq_len, int):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: when would this be true / false?

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -602,26 +620,45 @@ def __call__(
         # local seq=0 and breaking GradAccumulator's slice writes.
         # ``local_map`` runs the chunking body on plain tensors; under the
         # non-DTensor (eager) path we call ``_chunk_local`` directly.
-        # Equal chunk sizes also match GradAccumulator's sequential slice
-        # writes, which use one chunk length for each write offset.
+        def _chunk_count(seq_len: int | torch.SymInt) -> int:
+            from torch.fx.experimental.symbolic_shapes import optimization_hint
+
+            seq_len_hint = optimization_hint(seq_len)
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: Didn't understand why we need this, could you say a bit more?

btw I don't think we need to handle the uneven case where `chunk_len` doesn't divide local `seq len`

**Comment by felipemello1** [comment]
oh damn, Sol did me dirty. I asked it to only push to my fork! hahaha. The PR is not ready for review. Sorry @tianyu-l 

Its complete slop as it is. The gains are real though. MFU 32 -> 38%,  TPS 3,072 -> 3,600.

Its honestly unclear to me where the gains come from. The happen both when we compile and when we increase num chunks in eager. I need to investigate the exact mechanism a bit better.
